### ISSUE TITLE
_version not supported in elasticsearch 6.1 onwards

### ISSUE THREAD

**Issue by babadofar** [confluentinc/kafka-connect-elasticsearch#295]
When using the option "key.ignore":false, a **_version** field is added to the bulk index metadata.  
This will not work in Elasticsearch versions from 6.1 onwards, I tested it in Elasticsearch,  in 6.40 and 7.0-rc1.  The field should be named **version**, without the leading underscore

https://www.elastic.co/guide/en/elasticsearch/reference/6.1/docs-bulk.html#bulk-versioning

`WorkerTask.java:219)\n\tat java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)\n\tat java.util.concurrent.FutureTask.run(FutureTask.java:266)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n\tat java.lang.Thread.run(Thread.java:748)\nCaused by: org.apache.kafka.connect.errors.ConnectException: Bulk request failed: {\"root_cause\":[{\"type\":\"illegal_argument_exception\",\"reason\":\"Action/metadata line [1] contains an unknown parameter [_version]\"}],\"type\":\"illegal_argument_exception\",\"reason\":\"Action/metadata line [1] contains an unknown parameter [_version]\"}\n\tat io.confluent.connect.elasticsearch.bulk.BulkProcessor$BulkTask.handleMalformedDoc(BulkProcessor.java:428)\n\tat io.confluent.connect.elasticsearch.bulk.BulkProcessor$BulkTask.execute(BulkProcessor.java:382)\n\tat io.confluent.connect.elasticsearch.bulk.BulkProcessor$BulkTask.call(BulkProcessor.java:343)\n\tat io.confluent.connect.elasticsearch.bulk.BulkProcessor$BulkTask.call(BulkProcessor.java:329)`

**Comment by creativelikeadog** [comment]
Having the same problem, let me know if there is any workaround.

**Comment by jpeyper** [comment]
What version of the connector are you running?

Just did a quick test with ES 6.7.1 and KCES 5.2.1 as this would have been a huge blocker for me and everything worked fine. 

On a surface level, it seems to be setting the properly correctly on the master branch - https://github.com/confluentinc/kafka-connect-elasticsearch/blob/master/src/main/java/io/confluent/connect/elasticsearch/jest/JestElasticsearchClient.java#L364

**Comment by babadofar** [comment]
I'm on kafka connect  2.1.1-cp1 , elasticseach 6.7.1, and the key.ignore: false option works, but the logs are flooded with warnings about the _version field begin deprecated.  Most likely this will not work  in elastisearch 7. 
My guess is that this is due to the badly outdated versions of Jest which is used by the connector - , 2.4.0, which matches elasticsearch version 2.

**Comment by scooterGrun** [comment]
This repository contains integration tests against ES 2.4.1 and uses Jest 2.4.0, see [https://github.com/searchbox-io/Jest/blob/master/README.md#Compatibility](https://github.com/searchbox-io/Jest/blob/master/README.md#Compatibility)

Given ES 5 is already end of life ([https://www.elastic.co/support/eol](https://www.elastic.co/support/eol)), I don't see any reason to continue to support the older versions ES <= 5.

**Comment by SunilDhaker** [comment]
Any updates on ES 6.1 + support ?

**Comment by talaloni2** [comment]
Hi, I want to contribute to this project and use the ElasticsearchRestHighLevelClient instead of Jest client(For versions 6.5 and on).
Before I do this, I want to know if anybody had already tried to replace the whole jest implementation in the past.
Would be happy to hear an advice from someone who had experience with this project in order to prevent problems during development and merge.

**Comment by talaloni2** [comment]
Changed jest version to 6.3.1. works perfect for me on elasticsearch 6.5.4
Created pull request [https://github.com/confluentinc/kafka-connect-elasticsearch/pull/318](url)

**Comment by nabilschear** [comment]
I"m also seeing this in my logs like @babadofar 
`elasticsearch_1  | [2019-06-21T01:38:04,523][WARN ][o.e.d.c.ParseField       ] [KYoUJrT] Deprecated field [_version_type] used, expected [version_type] instead`

this is with elastic 6.8.1 and the kafka-connector 5.2.1 and key.ignore set to false.

is upgrading jest as easy as replacing the jar @talaloni2 ?

**Comment by talaloni2** [comment]
> I"m also seeing this in my logs like @babadofar
> `elasticsearch_1 | [2019-06-21T01:38:04,523][WARN ][o.e.d.c.ParseField ] [KYoUJrT] Deprecated field [_version_type] used, expected [version_type] instead`
> 
> this is with elastic 6.8.1 and the kafka-connector 5.2.1 and key.ignore set to false.
> 
> is upgrading jest as easy as replacing the jar @talaloni2 ?

I've already changed the jest version and fixed the errors caused by it. As for now, I am waiting for my pull request to be merged.

**Comment by mattferroni** [comment]
hey @talaloni2 , thanks for your fix, it worked for me, let's wait for the pull request to be merged...

In the meanwhile, a hint for those that need to deploy this connector in a Docker container: 

- build a **fat jar** from the code of the [merge request](https://github.com/confluentinc/kafka-connect-elasticsearch/pull/318), in order to bring the right dependencies with you;
- use the last official Docker image available [confluentinc/cp-kafka-connect:5.2.2](https://hub.docker.com/r/confluentinc/cp-kafka-connect/) and load the jar as an external plugin.

Using a non-fat jar was not working for me, probably because of version mismatch (5.2.2 vs. cp 5.4.0): this seems the fastest way to make it work. :)

**Comment by talaloni2** [comment]
> hey @talaloni2 , thanks for your fix, it worked for me, let's wait for the pull request to be merged...
> 
> In the meanwhile, a hint for those that need to deploy this connector in a Docker container:
> 
> * build a **fat jar** from the code of the [merge request](https://github.com/confluentinc/kafka-connect-elasticsearch/pull/318), in order to bring the right dependencies with you;
> * use the last official Docker image available [confluentinc/cp-kafka-connect:5.2.2](https://hub.docker.com/r/confluentinc/cp-kafka-connect/) and load the jar as an external plugin.
> 
> Using a non-fat jar was not working for me, probably because of version mismatch (5.2.2 vs. cp 5.4.0): this seems the fastest way to make it work. :)

If you are using docker compose, this might work, but if you plan to deploy on k8s, this might be a bit more problematic to automate it. 
An advice for those who need to deploy this image for now is to build an actual image using the latest kafka connect version and override the kafka-connect-elasticsearch directory with the mvn package result directory(worked for me using the complied directory packaged by the code in the pull request I've created)

**Comment by klalafaryan** [comment]
I am getting still the same error:
```
2019-08-01 16:50:53,372 ERROR  ||  Failed to execute batch 35 of 1 records after total of 6 attempt(s)   [io.confluent.connect.elasticsearch.bulk.BulkProcessor]
org.apache.kafka.connect.errors.ConnectException: Bulk request failed: {"root_cause":[{"type":"illegal_argument_exception","reason":"Action/metadata line [1] contains an unknown parameter [_version]"}],"type":"illegal_argument_exception","reason":"Action/metadata line [1] contains an unknown parameter [_version]"}
	at io.confluent.connect.elasticsearch.bulk.BulkProcessor$BulkTask.handleMalformedDoc(BulkProcessor.java:428)
	at io.confluent.connect.elasticsearch.bulk.BulkProcessor$BulkTask.execute(BulkProcessor.java:382)
	at io.confluent.connect.elasticsearch.bulk.BulkProcessor$BulkTask.call(BulkProcessor.java:343)
	at io.confluent.connect.elasticsearch.bulk.BulkProcessor$BulkTask.call(BulkProcessor.java:329)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```
`kafka-connect-elastic` version 5.3.0 
`elastic-search`  version 7.2.0

**Comment by dkirrane** [comment]
Question: Does the merged Elasticsearch 7 PR https://github.com/confluentinc/kafka-connect-elasticsearch/pull/327 supersede this?

**Comment by levzem** [comment]
this issue is solved by #327 and thus will be closing this issue.

Please re-open if there are any further questions
